### DISCUSSION TITLE
Importing a `js` file could fall back to `ts` file in a `ts` file

### DISCUSSION THREAD

**Discussion by Conaclos** [denoland/deno#18293] [Ideas]
Hi!

One thing that prevents me from using _Deno_ for my TypeScript projects is the incompatibility of Deno with other tools. in particular _ESbuild_ and _TSC_.

I use _ESbuild_ and _TSC_ to transpile my projects for npm and the browser. ESbuild and TSC do not change the extension of import paths. That's mean we have to use js extension to import a TypeScript file.

It could be nice to implement a very simple resolution strategy for _Deno_: in a TypeScript file (file ending with ts extension): an import path ending with a _js_ extension is resolved to the same file, but ending with a ts extension if the _js_ file does not exist.

```ts
// a.ts
import * as b from "./b.js" // resolves to "./b.js" or "./b.ts" if "./b.js" does not exist
...
```

**Comment by lilnasy** [comment]
This is a great suggestion, but it probably won't receive as much attention as an issue would.

**Reply by Conaclos** [reply]
I opened an issue #18390.

**Comment by Conaclos** [comment]
@lilnasy 
I can transfer the discussion to an issue. However, I do not know if it is ok to post a feature request in the issue tricker...

**Reply by lilnasy** [reply]
It happens all the time. All the discussion regarding new and planned features takes place in issues.

**Comment by nayeemrmn** [comment]
Could you elaborate on the use case further? What need is there run those build tools as opposed to ones which _do_ work with deno code, like https://deno.land/x/esbuild@v0.17.12?

**Reply by lilnasy** [reply]
There are plenty of libraries written for node that are compatible with deno except for their use of .js extension. If deno resolves the specifiers to .ts files, those libraries can be used in deno without any modification.

**Reply by nayeemrmn** [reply]
Yeah but they only run on node by transpiling them and publishing that on npm. ~~Worst case targeting deno without modification requires the same thing with a different transpiler+registry~~ but nowadays deno can use those with npm imports. That's less work for everyone than dual publishing, and using the compat mechanism is a reasonable ask wrt 'libraries written for node'.

**Reply by Conaclos** [reply]
Given the following files:

```ts
// @filname a.ts
export * from "./b.ts";

// @filname b.ts
export const X = 1;
```

And running the following command on node (or using the JS API in node or deno):

```sh
esbuild *.ts --outdir=dist
```

We got the following outputs:

```ts
// @filname dist/a.js
export * from "./b.ts";

// @filname dist/b.js
export const X = 1;
```

Note that the file path are not rewritten by _ESbuild_. _TSC_ produce the same output.
This makes the transpiled code impossible to run on deno, node, and the browser.

TSC and ESbuild supports js extension that resolve to ts extension.
Then the following code is valid:

```ts
// @filname a.ts
export * from "./b.js";

// @filname b.ts
export const X = 1;
```

Making the previous command outputs the following files:


```ts
// @filname a.js
export * from "./b.js";

// @filname b.js
export const X = 1;
```
